### ISSUE TITLE
[AMD][kernel][Tier 6] Port sol-execbench linear_backward (2× matmul bwd, fp32) to FlyDSL

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#40]
# FlyDSL implementation: port sol-execbench's linear_backward problem to FlyDSL

Cross-references:
- `amdpilot-org/SOL-ExecBench` — AMD fork of `nvidia/sol-execbench` (the GPU kernel eval framework that scores AI-generated kernel solutions for correctness + performance against a PyTorch reference). Currently NVIDIA-only: CUDA / Triton / CUTLASS / CuTe DSL. Zero AMD support.
- `examples/pytorch/linear_backward/` in that fork — the backward pass of a `nn.Linear` layer. The `definition.json` names it `linear_backward`; `op_type: "backward"`. The reference is two matmuls: `grad_input = grad_output @ weight` and `grad_weight = grad_output.mT @ x`.
- Sibling FlyDSL sol-execbench issue: #38 (rmsnorm, Tier 4) — the first suite-coverage target. This issue is the **second** coverage target and the **first backward-pass kernel** in the suite. FlyDSL's backward support should be validated here.
- This issue depends on the **gemm port** (issue being filed in parallel) for the forward matmul API patterns — `grad_input` and `grad_weight` are both GEMMs (one with transposed-B, one with transposed-A) and should reuse the same `compile_*gemm*` builder family as the forward gemm port.
- Production-hotspot-driven FlyDSL backward issues: #31 (FA-2 bwd), #32 (MoE bwd grouped-GEMM), #34/#35 (MoE bwd dW / dActivation), #36/#37 (fused linear-head + CE bwd). Those are training-pipeline-driven. This issue is **suite-coverage-driven**: the deliverable is "make FlyDSL a scoreable solution language for a sol-execbench backward problem."

## Why this issue exists

sol-execbench scores AI-generated kernel solutions across four NVIDIA-only
solution languages (PyTorch/Triton/CUTLASS/CuTe DSL). For the AMD port we
need FlyDSL to be a scoreable solution language for **backward** problems
too, not just forward. The `linear_backward` problem is the right first
backward target because:

- It is the simplest backward kernel in the suite: two plain matmuls, no
  fusion, no activation, no reduction. `grad_input = grad_output @ weight`
  (M×N @ N×K → M×K, transposed-B GEMM) and `grad_weight = grad_output.mT @ x`
  (N×M @ M×K → N×K, transposed-A GEMM).
- The reference is plain PyTorch (`torch.matmul`), so the correctness
  oracle is unambiguous. The original solution is pure PyTorch
  (`solution_python.json` — `@torch.no_grad()` + two `@` ops), so there is
  no Triton baseline to beat; the performance bar is **PyTorch eager on
  AMD with `torch.compile` off**.
- FlyDSL already ships a production GEMM family
  (`kernels/preshuffle_gemm.py`, `kernels/hgemm_splitk.py`,
  `kernels/rdna_f16_gemm.py`, `kernels/gemm_common_gfx1250.py`) covering
  fp8/fp4/bf16/fp16 on gfx942/gfx950/gfx1250. The work here is **not**
  "write a new GEMM from scratch" — it is "wire two existing FlyDSL GEMM
  builders into the sol-execbench backward solution contract, prove
  parity vs the PyTorch reference, and benchmark vs the PyTorch eager
  baseline." The fp32 dtype is the only gap — FlyDSL's GEMM family
  currently centers on bf16/fp16/fp8/fp4; a fp32 path must be wired
  through (likely via the `hgemm_splitk` builder extended to f32, or a
  dedicated f32 GEMM).
- This is the **first backward-pass kernel** in the sol-execbench →
  FlyDSL coverage sweep. FlyDSL's backward support (transposed-A and
  transposed-B GEMM via layout algebra — `composition` / `divide` on
  the K-mode) should be validated here, on the simplest possible
  backward shape, before tackling fused or MoE backward problems.

## Goal (one sentence)

Implement, unit-test, and benchmark a FlyDSL `linear_backward` kernel
that satisfies the `linear_backward` definition from
`amdpilot-org/SOL-ExecBench`, and report a clear pass/fail vs the
PyTorch eager baseline on both MI300X (gfx942) and MI355X (gfx950).

## Hardware / workload

```text
Hardware:  1 × MI300X (gfx942) + 1 × MI355X (gfx950)
Models:    (none — this is a kernel benchmark, not an LLM workload)
Shape:     grad_output [M, N] float32
           x            [M, K] float32
           weight       [N, K] float32
           grad_input   [M, K] float32   (= grad_output @ weight)
           grad_weight  [N, K] float32   (= grad_output.mT @ x)
           (M, N, K) varies per workload.jsonl — 3 shapes shipped:
           (64, 64, 128), (128, 128, 256), (256, 256, 512)
Baseline:  sol-execbench PyTorch reference on AMD (torch.compile off)
           run via the unified aiter+triton+flydsl Docker image
Target:    max_abs_diff < 1e-2 vs PyTorch fp32 reference on all tested shapes
           kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
```

The workload set (`workload.jsonl`) ships 3 concrete `(M, N, K)` shapes:
(64, 64, 128), (128, 128, 256), (256, 256, 512). All three are
square-ish small-MNK regimes — bandwidth-bound, not compute-bound. The
parity/bench harness should cover all 3 shipped shapes at minimum, plus
at least one larger shape (e.g. (1024, 1024, 1024)) to exercise the
compute-bound regime where FlyDSL's tiled MFMA pipeline should win.

## Scope

1. `kernels/sol_execbench/linear_backward.py` — FlyDSL kernel
   implementing the backward pass of `nn.Linear` as two GEMMs:
   - `grad_input  = grad_output @ weight`       (M×N @ N×K → M×K, transposed-B)
   - `grad_weight = grad_output.mT @ x`         (N×M @ M×K → N×K, transposed-A)
   Reuse the existing `kernels/hgemm_splitk.py::compile_hgemm_kernel`
   builder (extended to fp32 if needed) or `kernels/preshuffle_gemm.py`
   for the transposed-A/transposed-B variants — the new file is the
   sol-execbench-facing wrapper (entry point + `run()` returning both
   grads), not a reimplementation. The transposed-A and transposed-B
   layouts should be expressed via FlyDSL layout algebra
   (`composition` / `divide` on the K-mode), not via a manual
   `torch.transpose` in the wrapper.
2. `tests/kernels/test_sol_execbench_linear_backward.py` — parity
   harness vs the fp32 PyTorch reference from `definition.json`. Tests
   multiple `(M, N, K)` values: the 3 shipped shapes plus (1024, 1024,
   1024). Random inputs at standard normal scale (matching
   `generate_inputs` — divided by `N**0.5` / `K**0.5`),
   `torch.manual_seed(0)` for determinism. Checks both `grad_input` and
   `grad_weight` outputs.
3. `tests/kernels/bench_sol_execbench_linear_backward.py` — reports
   `kernel_ms` per shape + comparison vs the PyTorch eager baseline
   (run via the unified Docker image, same shape, same hardware, same
   fp32 dtype, `torch.compile` off).
4. `assessment.md` (≤300 words) — ranking which shapes FlyDSL
   wins/ties/loses vs PyTorch eager, with a one-paragraph hypothesis
   for the loss shapes (e.g. small-MNK is bandwidth-bound and the
   two-kernel launch overhead dominates; large-MNK is compute-bound
   and FlyDSL's tiled MFMA pipeline should win).

## Acceptance

- **Correctness:** `max_abs_diff < 1e-2` vs the PyTorch fp32 reference
  on all tested `(M, N, K)` values (the 3 shipped shapes + (1024, 1024,
  1024)), for **both** `grad_input` and `grad_weight` outputs. This
  matches sol-execbench's default `max_atol=1e-2` from
  `docs/workload.md`.
- **Performance:** `kernel_ms < 0.9 × PyTorch_eager_baseline` on ≥80%
  of the tested shapes, on at least one of MI300X / MI355X.
- **Coverage:** works on both gfx942 (MI300X) and gfx950 (MI355X). The
  fp32 GEMM path must exercise both architectures — if the existing
  `hgemm_splitk` builder only covers gfx942, a gfx950 path must be
  added (likely via `preshuffle_gemm` with fp32 in/out dtype).

## Out of scope

- **Optimizer step / SGD / Adam update** — this is the backward pass
  only (grad_input + grad_weight). The optimizer step is a separate
  elementwise problem and is not in sol-execbench's linear_backward
  definition.
- **Weight decay / L2 regularization** — not in the reference; the
  reference is plain `grad_output @ weight` and `grad_output.mT @ x`.
- **Gradient accumulation across microbatches** — the reference is a
  single backward pass with no accumulation buffer; accumulation is a
  training-loop concern, not a kernel concern.
- **Bias gradient** — `nn.Linear` with `bias=False` (the definition
  has no bias input); `grad_bias` would be a reduction, not a GEMM,
  and is not in scope.
- **Fused forward + backward** — sol-execbench's linear_backward is
  backward-only. A fused fwd+bwd linear layer is a separate, harder
  problem and gets its own issue.
- **Other sol-execbench problems** — `gemma3_swiglu` and the rest of
  the suite are follow-ups. This issue is the **first backward**
  coverage target, not a sweep.
- **Modifying the sol-execbench eval harness** — the wrapper conforms
  to the existing solution contract (`entry_point`, `destination_passing_style`,
  `dependencies`); no harness changes.

## Reproduction

To run the PyTorch eager baseline locally:

```bash
# 1. Build the unified aiter+triton+flydsl Docker image (not yet built;
#    Dockerfile at /home/xiasun/unified-rocm.Dockerfile on the AMDPilot
#    fleet — ROCm 7.x base, Triton + AIter + FlyDSL installed side-by-side).
docker build -f /home/xiasun/unified-rocm.Dockerfile -t unified-rocm:latest .

# 2. Clone the AMD fork and run the sol-execbench driver on the
#    linear_backward problem with the PyTorch solution.
git clone https://github.com/amdpilot-org/SOL-ExecBench
cd SOL-ExecBench/examples/pytorch/linear_backward
# (driver invocation TBD — depends on the sol-execbench runner CLI,
#  which is being wired up in parallel to this issue)
```

The FlyDSL solution is benchmarked by pointing the same driver at
`kernels/sol_execbench/linear_backward.py` instead of `kernel.py`.

## Definition of done

- [ ] FlyDSL kernel implemented under
      `kernels/sol_execbench/linear_backward.py` (two GEMMs via
      `compile_hgemm_kernel` or `preshuffle_gemm`, transposed-A and
      transposed-B expressed via layout algebra).
- [ ] Parity test passes on MI300X (gfx942): `max_abs_diff < 1e-2` on
      all tested `(M, N, K)` values for both `grad_input` and
      `grad_weight`, deterministic on rerun.
- [ ] Parity test passes on MI355X (gfx950): same bound, exercising
      the gfx950 fp32 GEMM path.
- [ ] Benchmark shows `kernel_ms < 0.9 × PyTorch_eager_baseline` on
      ≥80% of the tested shapes, on at least one of MI300X / MI355X.
- [ ] `assessment.md` (≤300 words) submitted with per-shape
      win/tie/loss table + hypothesis for losses.
- [ ] Issue closed with a one-line summary linking to the kernel, the
      test, the benchmark, and the assessment; names the next
      sol-execbench problem to port as a follow-up issue.

## ## Leaderboard baseline to beat

These are the current best public results on NVIDIA B200 (from the
[sol-execbench leaderboard](https://research.nvidia.com/benchmarks/sol-execbench/leaderboard))
for problem `090_batched_2d_rope_position_encoding_backward` (kernel_id=90). The agent's
FlyDSL kernel on MI300X/MI355X should aim to beat the **Scoring Baseline**
(optimized PyTorch) and approach the **top human/AI submissions**, normalized
by cross-arch roofline.

```text
GPU:           NVIDIA B200 (leaderboard) vs AMD MI300X/MI355X (this task)
Latency unit:  ms

SOL Bound (theoretical roofline):     0.001736 ms   (SOL-Score=1.0, speedup=28.69×)
Scoring Baseline (optimized PyTorch): 0.035036 ms   (SOL-Score=0.4946, speedup=0.98×)
Reference Implementation (plain PyTorch): 0.063860 ms   (SOL-Score=0.3500, speedup=0.60×)

Top public submissions:
  - **Rank 1** (`Recursive`): 0.005363 ms, SOL-Score=0.8942, speedup=7.42× vs reference
  - **Rank 2** (`doubleAI`): 0.005365 ms, SOL-Score=0.8870
  - **Rank 3** (`Qin Liu @ UCD`): 0.005490 ms, SOL-Score=0.8851
```

**Cross-arch note:** B200 timings are not directly comparable to MI300X/MI355X
(different GPU arch, different roofline). The agent should:
1. Measure the PyTorch reference on MI300X/MI355X (same shapes, same Docker image)
2. Compute the FlyDSL kernel's speedup vs the on-AMD PyTorch reference
3. Compare that speedup ratio to the leaderboard's `avg_speedup` ratios above
4. Target: FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × (Rank-1 speedup vs PyTorch-reference-on-B200)

## Definition of done

- [ ] FlyDSL kernel passes correctness (max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes)
- [ ] FlyDSL kernel_ms measured on MI300X (gfx942) AND MI355X (gfx950) across >=5 workload shapes
- [ ] FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × Rank-1 speedup vs PyTorch-reference-on-B200 (=7.42× if top1 exists)
- [ ] `assessment.md` (<=300 words) ranks FlyDSL vs PyTorch eager on each shape, identifies win/tie/lose regimes
- [ ] Patch modifies actual kernel source (not just bench scripts) — bench-only patches are REJECTED

## Non-goals / rejection criteria

- Bench-only patches (no FlyDSL kernel source changes) are rejected
- Patches that set expected_value = measured_value (tautological pass) are rejected
- The metric contract must compare FlyDSL vs PyTorch reference on AMD, not an absolute ms
- "Completed" without a measured speedup ratio vs the on-AMD PyTorch reference is rejected

**Comment by sunxxuns** [comment]
## AMDPilot acceptance target

**The deliverable is a FlyDSL kernel.** A PyTorch reimplementation of this
problem -- with or without `torch.compile` -- is the *baseline*, not a result.
It does not close this issue no matter how closely it matches the reference.

### Implementation contract (hard requirement)

The submitted kernel MUST be written in FlyDSL and go through the FlyDSL
compiler. Concretely, the kernel module has to actually use the DSL, e.g.:

```python
import flydsl.compiler as flyc
from flydsl.expr import arith as ea, gpu, vector as ev
```

and dispatch through `flyc.jit` / `flyc.kernel`. A file placed under
`FlyDSL/kernels/sol_execbench/` that only calls `torch.*` ops does NOT satisfy
this issue, even if it is numerically perfect and even if it is faster.

If you conclude the port is not expressible in FlyDSL today, say so explicitly
and report exactly which DSL capability is missing. That is a legitimate,
useful outcome. Silently substituting a PyTorch implementation is not.

### Baseline definition

The baseline is the sol-execbench PyTorch reference (`reference.py`) running
eager, with `torch.compile` OFF. Measure it yourself in this run, on the same
GPU and the same workload shapes you benchmark your kernel on. Do not carry a
number over from another run, another GPU, or the issue text.

### Performance target

Your FlyDSL kernel must be **at least 10% faster than that measured baseline**
-- `kernel_ms <= 0.90 x baseline_ms` -- on **at least 80% of the tested
workload shapes**.

### Correctness target

`max_abs_diff < 1e-2` versus the PyTorch reference on **all** tested shapes. A
faster kernel that fails this is a failure, not a tradeoff.

### What you must report

Report a per-shape table containing, for every tested shape: the measured
eager baseline ms, your FlyDSL kernel ms, the ratio, and win/loss. Then state
the win rate against the 80% bar. A result without the baseline column next to
the kernel column is not checkable and will not be credited.
